crates/libcontainer/src/process/container_init_process.rs

```diff
@@ -270,6 +270,68 @@ fn reopen_dev_null() -> Result<()> {
     Ok(())
 }
 
+fn unmount_or_hide(syscall: &dyn Syscall, target: impl AsRef<Path>) -> Result<()> {
+    let target_path = target.as_ref();
+    match syscall.umount2(target_path, MntFlags::MNT_DETACH) {
+        Ok(_) => Ok(()),
+        Err(SyscallError::Nix(nix::errno::Errno::EINVAL)) => syscall
+            .mount(
+                None,
+                target_path,
+                Some("tmpfs"),
+                MsFlags::MS_RDONLY,
+                Some("size=0k"),
+            )
+            .map_err(InitProcessError::SyscallOther),
+        Err(err) => Err(InitProcessError::SyscallOther(err)),
+    }
+}
+
+fn move_root(syscall: &dyn Syscall, rootfs: &Path) -> Result<()> {
+    unistd::chdir(rootfs).map_err(InitProcessError::NixOther)?;
+    unmount_or_hide(syscall, "/sys")?;
+    unmount_or_hide(syscall, "/proc")?;
```

Comment:
ya, the logic is borrowed from crun and similar logic is implement in runc as well https://github.com/containers/crun/blob/53cd1c1c697d7351d0cad23708d29bf4a7980a3a/src/libcrun/linux.c#L2780
the reason behind these is to umount or hide the proc and sys fs before change it's root to new rootfs, umount here is safe because procfs and sysfs is privately mounted in the containers mountns, this is not required when change rootfs with pivotroot as the old rootfs will be swap to different location